Adding Go and C# .Net AMQP Samples

amantewari154 opened the pull request:
New directory "amqp" created in the root of the project.
Directories /amqp/Go and /amqp/dotnet created along with the samples.
References to the code added in the main project Readme.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ibm-messaging/mq-dev-patterns?pullRequest=191) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ibm-messaging/mq-dev-patterns?pullRequest=191) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: shivaprasaddn<br/>:x: amantewari154<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ibm-messaging/mq-dev-patterns?pullRequest=191) it.</sub>

chughts commented on amqp/Go/README.md:
Please add which version of go this code has been tested with.

chughts commented on amqp/Go/README.md:
```diff
@@ -0,0 +1,56 @@
+# IBM MQ Go AMQP 1.0 samples
+The Go samples are based on go-amqp (a Go AMQP 1.0 client implementation) as provided here: [AMQP 1.0 client module for Go](https://github.com/Azure/go-amqp)
+## Go Samples
+
```
As the samples perform both put/get and pub/sub, a better name for them is amqpProducer.go and amqpConsumer.go

chughts commented on amqp/Go/amqpGet.go:
```diff
@@ -0,0 +1,181 @@
+var MQ_ENDPOINTS MQEndpoints
+
+func init() {
+    //Parsing the input
```
Break down the code into further methods / functions to make each logical unit of functionality easier to read.

chughts commented on amqp/Go/amqpPut.go:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    // Create a sender
```
Break down the code into methods / functions to make each unit of functionality easer to read.

chughts commented on amqp/README.md:
```diff
@@ -0,0 +1 @@
+
```
Remove this empty file

chughts commented on amqp/dotnet/SimplePut.cs:
```diff
@@ -0,0 +1,106 @@
+/*                                                                                      */
+/****************************************************************************************/
+
+
```
SimplePut and SimplePub have a lot of overlap that should be pushed into a shared class.

chughts commented on amqp/dotnet/SimpleGet.cs:
```diff
@@ -0,0 +1,117 @@
+
+namespace ibmmq_amqp_samples
+{
+    class SimpleGet
```
A lot of overlap between get and sub, could be consolidated into a common class.

chughts commented on amqp/dotnet/env.json:
```diff
@@ -0,0 +1,46 @@
+﻿{
+  "MQ_ENDPOINTS": [
+    {
```
strip the env.json to the endpoint that you are expecting to use.

shivaprasaddn replied:
Removed the redundant Endpoints.

amantewari154 commented:
Thank you for the thorough review. Made changes specific to Golang.
1. Added tested Go version to the Readme
2. Changed samples names to amqpProducer.go and amqpConsumer.go
3. Removed empty file amqp/README.md
4. Rewrote the applications to include methods/functions for better readability.

shivaprasaddn commented:
Thank You for the thorough Review,

Made necessary changes to .NET samples
- Created a SharedClass.cs file and added Put and Get messages to reuse.
- Changed other samples with respect to the SharedClass.cs
- Removed Redundant Endpoints in json file.

Please let me know if any changes required.

rcoppen commented on amqp/Go/amqpConsumer.go:
```diff
@@ -0,0 +1,215 @@
+/*  FILE NAME:      amqpConsumer.go                                                     */
+/*                                                                                      */
+/*  DESCRIPTION:    A Go applicaton used to receive messages from broker-IBM MQ.        */
+/*  AMQP 1.0 in use. Go-azure library is used.                                          */
```
```suggestion
/*  AMQP 1.0 in use. An Azure go-amqp library is used.                                          */

```

rcoppen commented on amqp/Go/amqpConsumer.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+func receiveMessages(receiver *amqp.Receiver) error{
+    for{
```
This looks like a tight get loop. Is it possible for the client to wait for a period of time for a message to appear?

amantewari154 replied:
By default, if there is no activity, ie. no frames being exchanged between the server and client, the connection will be forcibly closed in 15 seconds. This is due to the timeout negotiated when the connection was established. This can be modified by providing ConnOptions (https://pkg.go.dev/github.com/Azure/go-amqp#ConnOptions).

So, in a pub/sub case too, the same timeout will be considered and the publisher should begin to send frames before it expires.

rcoppen commented on amqp/Go/amqpProducer.go:
```diff
@@ -0,0 +1,218 @@
+/*  FILE NAME:      amqpProducer.go                                                     */
+/*                                                                                      */
+/*  DESCRIPTION:    A Go applicaton used to send 10 messages to broker-IBM MQ.          */
+/*  AMQP 1.0 in use. Go-azure library is used.                                          */
```
```suggestion
/*  AMQP 1.0 in use. An Azure go-amqp library is used.                                          */

```

rcoppen commented on amqp/Go/amqpProducer.go:
```diff
@@ -0,0 +1,218 @@
+
+    // Send the messages
+    for i:=0; i<10; i++ {
+        err = sendMessages(sender, msg)
```
```suggestion
        err = sendMessage(sender, msg)

```

rcoppen commented on amqp/Go/amqpProducer.go:
```diff
@@ -0,0 +1,218 @@
+    return sender, nil
+}
+
+func sendMessages(sender *amqp.Sender, msg *amqp.Message) error{
```
```suggestion
func sendMessage(sender *amqp.Sender, msg *amqp.Message) error{

```

rcoppen commented on amqp/dotnet/Env.cs:
```diff
@@ -0,0 +1,185 @@
+                return string.Join(",", connList);
+            }
+
+            //return "192.168.252.1(1414),192.168.252.1(1416)";
```
Delete this comment or make it informational

rcoppen commented on amqp/dotnet/SharedClass.cs:
```diff
@@ -0,0 +1,144 @@
+            // Create ReceiverLink
+            ReceiverLink receiver = new ReceiverLink(session, "client", source, OnAttached);
+
+            while (true)
```
This looks like a tight loop. Is it possible for receiver to get with wait to allow time for a message to appear from a producer? Two instances of the application could be run individually.

shivaprasaddn replied:
The receiver will wait for 60 seconds, and if no action occurs within that time, the connection will be closed.

rcoppen commented on amqp/dotnet/SimplePub.cs:
```diff
@@ -0,0 +1,56 @@
+                sharedClass.Password = env.Conn.app_password;
+                sharedClass.symbolName = env.Conn.topic_name;
+
+                sharedClass.PutMessages("topic");
```
The verbs are mixed up using this approach with a shared class. A Publish operation is a call a method called Put which suggests queues...

rcoppen commented on amqp/dotnet/SimpleSub.cs:
```diff
@@ -0,0 +1,55 @@
+                sharedClass.Password = env.Conn.app_password;
+                sharedClass.symbolName = env.Conn.topic_name;
+
+                sharedClass.GetMessages("topic");
```
The subscribe operation is 'getting' - I think the methods should be renamed to produce and consume to avoid confusion.

shivaprasaddn replied:
changed method names from PutMessages to Produce and GetMessages to Consume.

rcoppen reviewed (approved):
@amantewari154 thanks for making the changes raised by me and @chughts. I'll raise an issue for you to add the link to the API default timeouts into the README so that we signpost to users.
